programs/survival/horse_buddy.sc

```diff
@@ -0,0 +1,306 @@
+);
+
+__on_player_interacts_with_entity(p, horse, hand) -> (
+  kind = horse~'type';
+  if(kind == 'horse' || kind == 'donkey',
+    _horse_logic(p, horse);
+  );
+);
+
+
+//--- LOGIC n stuff ---//
+
+_horse_logic(p, horse) -> (
+
+    // reset training for new horses
+    if(horse~'uuid' != global_training:'uuid',
+      _rest_training(horse~'uuid');
+    );
+
+
+    item = p~'holds';
+    // when you feed the demanding horse what it wants
+    if(item:0 == global_training:'food',
+
+      // Pick a new craving
+      global_training:'generic.max_health' += _food_strengths():(item:0);
+      global_training:'food' = _random_food();
+
+      particle('happy_villager', horse~'pos'+[0,1.5,0], 8, 0.5);
+
+      // once the horse is satisfied, re-roll some stats.
+      if(global_training:'generic.max_health' > 16,
+        _roll_new_stats(horse);
+        schedule(10, '_after_hint', p)
+      );
+
+    //elif make the horse complain about its food.
+    ,_food_strengths():(item:0) || !rand(8),
+      schedule(10, '_food_hint', p)
+
+    //elif give random hits about potions and things.
+    ,!rand(5), 
+      schedule(10, '_random_hint', p)
+
+    );
+);
+
+//--- ROLL NEW STATS ---//
+
+_roll_new_stats(horse) -> (
+  effect_mod = _effect_stat_mods(reduce(horse~'effect', _a += _:0, []));
+  attributes = horse~'attribute';
+
+  // these are the magic numbers that try to balance out the training
+  // we need to turn numbers like 500 and 6000 into reasonable stat modifiers no larger than 1.
+  global_training:'horse.jump_strength' = global_training:'horse.jump_strength' * 0.2 * 0.002;
+  global_training:'generic.movement_speed' = global_training:'horse.jump_strength' * 0.125 * 0.002;
+  global_training:'generic.max_health' = global_training:'horse.jump_strength' * 0.33 * 0.002;
+
+  // build the new horse attributes
+  new_attrs = [];
+  for(pairs(_horse_stat_range()),
+    new_attrs += {
+      'Name' -> _:0,
+      'Base' -> _clamp( _:1:0, _:1:1, attributes:(_:0) + _roll_stat_for(_:0, 
+        effect_mod:(_:0)+min(0.4,global_training:(_:0))
+      ))
+    }
+  );
+
+  // merge in the attributes and show some nice particales
+  modify(horse, 'nbt_merge', nbt({'Attributes' -> new_attrs}));
+  particle( 'glow', horse~'pos'+[0,1.8,0], 32, 1);
+  sound('minecraft:entity.zombie_villager.cure',horse~'pos');
+
+  // reset the horses training and remove the effects
+  _rest_training(horse~'uuid');
+  modify(horse, 'effect');
+);
+
+
+//--- MESSAGES ---//
+
+_after_hint(p) -> _rand_message(p,'roll.after', 2);
+_random_hint(p) -> _rand_message(p,'random.hint', 9);
+_food_hint(p) -> (
+  _rand_message(p, str('food.%s', global_training:'food'), 2);
+);
+_message(p,code) -> (
+  display_title(p, 'actionbar', format(split('\\|','iy ' + _i18n(code))))
+);
+_rand_message(p,code,num) -> (
+  _message(p,str('%s.%d', code, floor(rand(num + 0.9))));
+);
+
+
+//--- LOCALS ---//
+
+_i18n(code) -> ({
+  'en'->{
+    'random.hint.0'       -> 'giv\'me jumpy jump magic!',
```

Comment:
The text is adorable, and totally untranslatable into any of the languages I know. I love it.

Reply:
lol, it's not like it needs to be 1-1 translation! just like basically, horse puns and stuff, with whatever horses sound like on paper in that language. Every language has horse puns! probably.